data/src/components/atom.rs

```diff
@@ -0,0 +1,423 @@
+    }
+
+    /// Update the state value.
+    pub fn write(&mut self, value: T) {
+        M::write(self, value)
+    }
+}
+
+impl<T: 'static> Atom<T, Normal> {
+    /// Given a pointer to an [`Atom`] in [`Normal`] mode, returns the offset of the internal field
+    /// that holds the actual value.
+    pub const FIELD_OFFSET: usize = {
+        // This expression will fail to type check when the internal representation of
+        // `Atom` changes. E.g. when it is no longer `T`. This is important to ensure the offset
+        // computed below is sound. We're effectively bringing the type assumption closer to where
+        // the assumption is used.
+        let _sanity = |atom: Atom<T, Normal>| -> T { atom.atom };
+
+        std::mem::offset_of!(Self, atom)
+    };
+
+    /// Construct an [`Atom`] in [`Prove`] mode.
+    ///
+    /// The initial proof value is equal to that of the given [`Atom`] in [`Normal`] mode.
+    pub fn start_proof(&self) -> Atom<T, Prove<'_>> {
+        Atom {
+            atom: ProveImpl {
+                previous: Source::Borrowed(&self.atom),
+                current: None,
+                read: Cell::new(false),
+            },
+        }
+    }
+}
+
+impl<A, B, M, N> PartialEq<Atom<B, N>> for Atom<A, M>
+where
+    A: PartialEq<B> + 'static,
+    B: 'static,
+    M: AtomMode,
+    N: AtomMode,
+{
+    fn eq(&self, other: &Atom<B, N>) -> bool {
+        A::eq(self, other)
+    }
+}
+
+impl<A: PartialEq + 'static, M: AtomMode> Eq for Atom<A, M> {}
+
+impl<T: 'static, M: AtomMode> Deref for Atom<T, M> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        M::deref(self)
+    }
+}
+
+impl<T: Clone + 'static, M: AtomMode> DerefMut for Atom<T, M> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        M::deref_mut(self)
+    }
+}
+
+impl<T: Default + 'static, M: AtomMode> Default for Atom<T, M> {
+    fn default() -> Self {
+        M::new(T::default())
+    }
+}
+
+impl<T: Encode + 'static> Foldable<HashFold> for Atom<T, Normal> {
+    fn fold(&self, _builder: HashFold) -> Hash {
+        Hash::blake3_hash(&self.atom).expect("Hashing should not fail")
+    }
+}
+
+impl<'normal, T: Encode + 'static> Foldable<HashFold> for Atom<T, Prove<'normal>> {
+    fn fold(&self, _builder: HashFold) -> Hash {
+        let value = self
+            .atom
+            .current
+            .as_deref()
+            .unwrap_or_else(|| &self.atom.previous);
+        Hash::blake3_hash(value).expect("Hashing should not fail")
+    }
+}
+
+impl<T: Encode + 'static> Foldable<MerkleTreeFold> for Atom<T, Prove<'_>> {
+    fn fold(&self, _builder: MerkleTreeFold) -> MerkleTree {
+        let data = serialise(self.atom.previous.deref()).expect("Serialisation should not fail");
+
+        // Determine whether the value has been read or written during proof generation. If so, we
+        // must mark it as not blinded in the Merkle tree.
+        let access = self.atom.read.get() || self.atom.current.is_some();
```

Comment:
ditto - in general, if it is written to before reading then presumably it _can_ be blinded - but it must exist in the output proof. ie its parent cannot be blinded

(I'm guessing a future follow up?)